internal/controller/controller.go

```diff
@@ -28,39 +37,227 @@ type JSONStructOr struct {
 	return func(res http.ResponseWriter, req *http.Request) {
-		const op = "controller.CreateSHortURLHandler"
+		const op = "controller.DeleteURLSHandler"
+		// check the body
 		body, err := io.ReadAll(req.Body)
+		c.logger.Info("Request body in DeleteURLSHandler", zap.String("body: ", string(body)))
+		if err != nil || len(body) == 0 {
+			c.logger.Error("Failed to read request body", zap.Error(err))
+			res.WriteHeader(http.StatusUnauthorized) // changed from 400
+			return
+		}
+		c.logger.Info("DeleteURLSHandler", zap.String("Request body: ", string(body)), zap.String("url", req.URL.String()))
+
+		// retrieve token if any
+		c.logger.Info("[INFO]", zap.Any("retrieveUserIDFromToken", op))
+		userID := c.retrieveUserIDFromTokenString(req)
+		if userID == "-1" { // can't retrieve token => error
+			c.logger.Info("Request token issues", zap.String("token", string(body)))
+			res.WriteHeader(http.StatusUnauthorized)
+			return
+		}
+		// check if user exists TODO: del that or not
+		err = c.storage.CheckIfUserExists(req.Context(), userID)
+		if err != nil {
+			if errors.Is(err, merrors.ErrUserNotFound) {
+				c.logger.Info("User not found", zap.String("user_id", userID))
+				res.WriteHeader(http.StatusUnauthorized)
+				return
+			}
+			// handle other possible errors (unexpected ones)
+			c.logger.Error("Failed to check if user exists", zap.Error(err))
+			res.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		// Replace the body with a new reader after reading from the original to reuse it
+		req.Body = io.NopCloser(bytes.NewBuffer(body)) // TODO unmarshal body
+		var buf bytes.Buffer
+		// feed data from the body into the buffer
+		if _, err := buf.ReadFrom(req.Body); err != nil {
+			c.logger.Error("[ERROR]", zap.Error(err))
+			http.Error(res, http.StatusText(http.StatusBadRequest), http.StatusBadRequest) // TODO [MENTOR]: BadRequest or InternalServerError?
+			return
+		}
+		defer req.Body.Close() // TODO [MENTOR] I didn't assign it should I close it??
+		// deserialize (unmarshal) the slice of aliases to delete
+
+		var aliasesToDel []string // ids  - the slice of aliases to del
+		if err := json.Unmarshal(buf.Bytes(), &aliasesToDel); err != nil {
+			c.logger.Error("[ERROR]", zap.Error(err))
+			http.Error(res, err.Error(), http.StatusBadRequest) // TODO [MENTOR]: BadRequest or InternalServerError?
+			return
+		}
+
+		// В случае успешного приёма запроса хендлер должен возвращать HTTP-статус 202 Accepted.
+		//Фактический результат удаления может происходить позже — оповещать пользователя
+		//об успешности или неуспешности не нужно. => notify the user
+		res.Header().Set("Content-Type", "application/json")
+		res.WriteHeader(http.StatusAccepted) // TODO [MENTOR]: when it would be send ?? I have a goroutine here!
+		// TODO gone status!!!!
+		// c.wg.Add(1) // todo define wait group in main and bring it to controller
+		//
+		// fill the channel to delete urls
+		c.delJobs <- JobToDelete{UserID: userID, LsURL: aliasesToDel}
```

Comment:
```suggestion
		go func() {
			c.delJobs <- JobToDelete{UserID: userID, LsURL: aliasesToDel}
		}
```
Буферизованный канал здесь не приводит к точному следованию заданию. Достаточно сделать одновременно 101 запрос - и последний уже не будет асинхронным.

Reply:
При данном раскладе не будет ли завершена работа данной горутины раньше положенного срока? Ведь тут её работа зависит от горутины более высокго уровня - горутины самого хэндлера.

Reply:
Что именно подразумевается под ".. и последний уже не будет асинхронным"?

Reply:
Именно так. Запрос, который переполнит буфер, будет вынужден дожидаться возможности слить туда все свои переданные id. Если мы передадим их в горутину, то запрос завершится, не дожидаясь окончания ее работы, а она захлопнется только по завершении всех записанных в нее действий. Здесь только одна ловушка - в это горутину не следует передавать контекст запроса и пытаться им оперировать, так как он почти гарантированно протухнет к моменту, когда дело дойдет до записи в БД.